src/fidasim.f90

```diff
@@ -6020,6 +6139,153 @@ subroutine track(rin, vin, tracks, ntrack, los_intersect)
+
+    integer :: cc, i, ii, mind, ncross, ir, iz, iphi
+    logical :: in_plasma1, in_plasma2, in_plasma_tmp, los_inter
+    integer, dimension(1) :: minpos
+    integer, dimension(3) :: ind
+    real(Float64) :: t, dT, dt1, inv_50, dr, dz, dphi
+    real(Float64), dimension(3) :: ri, vn, ri_tmp, n0, nz, p
+    real(Float64), dimension(3) :: side_min, side_max, outer_circle
+    real(Float64), dimension(3) :: redge, rstep, tedge, zstep
+    real(Float64), dimension(6) :: dt_arr
+
+    integer, dimension(3) :: gdims
+
+    !!! Assume ri is in cyl coordinates and vn is in machine coordinates
+    vn = vin ;  ri = rin ; ntrack = 0
+
+    if(dot_product(vn,vn).eq.0.0) then
+        return
+    endif
+
+    dr = inter_grid%dr
+    dz = inter_grid%dz
+    rstep = (/dr, 0.d0, 0.d0/)
+    zstep = (/0.d0, dz, 0.d0/)
+    dphi = inter_grid%dphi
+    minpos = minloc(abs(inter_grid%r - ri(1)))
+    ir = minpos(1)
+    ind(1) = ir
+    minpos = minloc(abs(inter_grid%z - ri(2)))
+    iz = minpos(1)
+    ind(2) = iz
+    minpos = minloc(abs(inter_grid%phi - ri(3)))
+    iphi = minpos(1)
+    ind(3) = iphi
+    side_min = (/inter_grid%r(ir), inter_grid%z(iz), inter_grid%phi(iphi)/)
+    side_max = (/inter_grid%r(ir), inter_grid%z(iz), inter_grid%phi(iphi+1)/)
+    nz = (/0.d0, 1.d0, 0.d0/)
+
+    inv_50 = 1.0/50.0
+    cc=1
+    los_inter = .False.
+    tracks%time = 0.d0
+    tracks%flux = 0.d0
+    ncross = 0
+    call in_plasma(ri,in_plasma1)
+    track_loop: do i=1,beam_grid%ntrack
+        if(cc.gt.beam_grid%ntrack) exit track_loop
+
+        !!! Not sure how to handle this if statement. Maybe keys to
+        !!! SpectralChords structure, i.e., incorporate something else
+        !!! for the los in the interpolation grid.
+        if((igrid_spec_chords%inter(ind(1),ind(2),ind(3))%nchan.ne.0) &
+            .and.(.not.los_inter))then
+            los_inter = .True.
+        endif
+        !!! End
+        !!! I assumed that ri & p0 are in cyl, and vn & n are in mc
+        !!! Double check the coordinate system going in and out for all
+        !!! subroutines.
+        call line_plane_intersect(ri, vn, side_min, nz, p, t, cyl=.True.)
+        dt_arr(1) = t
+        call line_plane_intersect(ri, vn, side_max, nz, p, t, cyl=.True.)
+        dt_arr(2) = t
+        redge = side_min + rstep
+        tedge = side_min + zstep
+        call rz_normal_vector(side_min, redge, tedge, n0)
+        call line_plane_intersect(ri, vn, side_min, n0, p, t, cyl=.True.)
+        dt_arr(3) = t
+        redge = side_max + rstep
+        tedge = side_max + zstep
+        call rz_normal_vector(side_max, redge, tedge, n0)
+        call line_plane_intersect(ri, vn, side_max, n0, p, t, cyl=.True.)
+        dt_arr(4) = t
+        ! inner_circle = side_min
+        call line_circle_intersect(ri, vn, side_min, p, t)
+        dt_arr(5) = t
+        outer_circle = side_min + rstep
+        call line_circle_intersect(ri, vn, outer_circle, p, t)
+        dt_arr(6) = t
```

Comment:
You call `line_plane/circle_intersect` many times and each time you are doing the same conversion back to cylindrical coordinates. This is inefficient. I would just use `uvw` coordinates throughout making it simpler and more efficient.